src/lib.rs

```diff
@@ -73,50 +73,34 @@ pub fn run(config: &Config) -> Vec<ViolationWithContext> {
         |_dir_entry, language, mut perform_search| {
             let mut instantiated_per_file_rules: HashMap<RuleName, Box<dyn RuleInstancePerFile>> =
                 Default::default();
-            perform_search(Box::new(
-                |capture_info: &CaptureInfo, file_contents, path| {
-                    let (instantiated_rule, rule_listener_index) = aggregated_queries
-                        .get_rule_and_listener_index(language, capture_info.pattern_index);
-                    let mut query_match_context = QueryMatchContext::new(
-                        path,
-                        file_contents,
-                        instantiated_rule,
-                        config,
-                        language,
-                    );
-                    instantiated_per_file_rules
-                        .entry(instantiated_rule.meta.name.clone())
-                        .or_insert_with(|| {
-                            instantiated_rule
-                                .rule_instance
-                                .clone()
-                                .instantiate_per_file(&FileRunInfo {})
-                        })
-                        .on_query_match(
-                            rule_listener_index,
-                            capture_info.node,
-                            &mut query_match_context,
-                        );
-                    if let Some(violations) = query_match_context.violations.take() {
+            perform_search(Box::new(|query_match: &QueryMatch, file_contents, path| {
+                run_match(
```

Comment:
DRY'ed up the "what do we do once we get called by a `tree-sitter-grep` "per-match" callback" a fair amount